### PR TITLE
Shadow v4

### PR DESCRIPTION
**Pull request by sei-amellinger**

### CODE REVIEW

File: bin/jb_run_experiment [hunk 1]
```diff
@@ -196,6 +180,16 @@ def main():
             show_command=show_command,
         )
 
+    # Alternative method to generate rules.json using an attack config.
+    attack_mgr = jbfs.AttackManager(args.experimentName)
+    attack_path = attack_mgr.get_experiment_attack_file()
+    if not rules_path.exists() and attack_path.exists():
+        logger.info(f"rules.json in experiment {args.experimentName} not found.")
+        logger.info(f"Generating experiment rules.json file for {args.experimentName} using the attack config.")
+        jbworkflow.run_command(dependencies=[],
```
Comment (sei-amellinger) [inline bin/jb_run_experiment:189]: The attack script should move into bin, be in the setup.py and use the jb_xxx naming scheme.

File: juneberry/tensorflow/trainer.py [hunk 2]
```diff
@@ -125,6 +130,12 @@ def setup(self) -> None:
         np.random.seed(self.model_config.seed)
         tf.random.set_seed(self.model_config.seed)
 
+        # Configure tensorflow to progressively allocate GPU memory rather than fully
```
Comment (sei-amellinger) [inline juneberry/tensorflow/trainer.py:133]: We don't need this twice do we?

File: experiments/shadowExperiment/attack.json [hunk 3]
```diff
@@ -0,0 +1,59 @@
+{
+    "format_version": "0.1.0",
+    "models": {
+        "shadow": "model_tests/lfw_baselines/leinonet-5-d/",
```
Comment (sei-amellinger) [inline experiments/shadowExperiment/attack.json:4]: I don't have this model.

File: juneberry/schemas/attack_schema.json [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                "disjoint_args": { "type": "array", "items": { "type": "object" } },
+                "private_disjoint_args": { "type": "object" },
+                "private_subset_args": { "type": "object" },
+                "query_watermarks": { "$ref": "#/definitions/plugin" },
+                "superset_args": { "type": "array", "items": { "type": "object" } },
+                "training_watermarks": { "$ref": "#/definitions/plugin" }
+            }
+        }
+    },
+    "required": [ ]
```
Comment (sei-amellinger) [inline juneberry/schemas/attack_schema.json:46]: Is nothing actually required?

File: juneberry/filesystem.py [hunk 5]
```diff
@@ -194,40 +194,275 @@ def load_file(path: str):
...
+    def get_shadow_model_dir(self, model_num: int, disjoint=False):
+        """
+        :param model_num: Integer that identifies a particular shadow model subdirectory.
+        :param disjoint: Boolean that controls whether to return the 'superset' (default) or
+        'disjoint' version of the desired property.
+        :return: The path to a particular shadow model subdirectory within either the
+        'superset' (default) or 'disjoint' shadow model subdirectory.
+        """
+        model_dir = self.get_shadow_subdir(disjoint=disjoint) / str(model_num)
+        if not model_dir.exists():
```
Comment (sei-amellinger) [inline juneberry/filesystem.py:403]: Path calls should probably not have side effects like this. For example, cleaning code is going to create a directory as a side effect of checking to see if the file is there.

File: juneberry/schemas/attack_schema.json (hunk 4 above)
Comment (sei-nwinski) [reply juneberry/schemas/attack_schema.json]: I added the required properties in 5b9088d.

File: juneberry/filesystem.py (hunk 5 above)
Comment (sei-nwinski) [reply juneberry/filesystem.py]: Addressed in e61391f.

File: bin/jb_run_experiment (hunk 1 above)
Comment (sei-nwinski) [reply bin/jb_run_experiment]: It looks like this was addressed in b775c34.

**Review by sei-nwinski — APPROVED** [review]

**Pull request merged**